I3 gyro cov

LordBismaya opened the pull request:
Moved the gyroData to a new Dataset folder. The gyro Estimation function is also updated.

LordBismaya committed:
Functionalised the FastSLAM code. Still have to move them to the respective folders. Code works with different motion and measurement models.

LordBismaya committed:
Some new changes. yet To create a separate datesets directory.

LordBismaya committed:
Moved the dataset to a separate folder. minor changes to the estimation code

icolwell commented on matlab_simulation/02-examples_practice/gyroCovEstimation/gyroCovEstimation.m:
```diff
@@ -0,0 +1,90 @@
+tmax = 20; % Put Stop_time here in seconds
+
+%% Load Data and Initial Calculations
+load('gyroData.mat'); % Load the Gyro Dataset.
```
I ran the "MobileRoboticsSetup.m" and then ran this file. It produced the following error:
Error using load, Unable to read file 'gyroData.mat': no such file or directory.
Since moving the data was part of the issue description, I would suggest updating the "MobileRoboticsSetup.m" file to include the new directory you created.

stevenwaslander replied:
Update works.

icolwell commented on matlab_simulation/02-examples_practice/gyroCovEstimation/gyroCovEstimation.m:
```diff
@@ -0,0 +1,90 @@
+clear all; close all; clc;
+RAD2DEG=180/pi;
+
+tmin = 10; % Put Start_time here in seconds
```
The formatting in this document is not consistent. For example: notice the difference between line 14 and 16, there is differences in the whitespace around the "=" sign. Please refer to the MatLab style guide:
[https://sites.google.com/site/matlabstyleguidelines/layout-documentation/white-space#TOC-Surround-and-by-spaces](https://sites.google.com/site/matlabstyleguidelines/layout-documentation/white-space#TOC-Surround-and-by-spaces)

icolwell commented on matlab_simulation/02-examples_practice/gyroCovEstimation/gyroCovEstimation.m:
```diff
@@ -0,0 +1,90 @@
+%% A Matlab program to calculate the convariances from a Gyroscope Data.
+% It demonstrates the following aspects:
```
Did you forget to mention what aspects this example demonstrates?

icolwell commented on matlab_simulation/02-examples_practice/gyroCovEstimation/gyroCovEstimation.m:
```diff
@@ -0,0 +1,90 @@
+Gez = normpdf(xvec,mean(G_data(:,3)),sqrt(Sf_err(3,3)))*length(G_err(:,3))*binsizeZ;
+
+subplot(3,1,1);bar(Xex,Nex);hold on;plot(xvec,Gex,'r','LineWidth',2);
+title('Comparison of Error Data and Gaussian Approximation');xlabel('Error Value');ylabel('Histogram Count');
```
All of these plots have the same titles yet they are displaying different data, I would suggest giving them appropriate titles to reduce confusion.

icolwell commented on matlab_simulation/02-examples_practice/gyroCovEstimation/gyroCovEstimation.m:
```diff
@@ -0,0 +1,90 @@
+title('Comparison of Error Data and Gaussian Approximation');xlabel('Error Value');ylabel('Histogram Count');
+subplot(3,1,2);bar(Xey,Ney);hold on;plot(xvec,Gey,'r', 'LineWidth',2);
+title('Comparison of Error Data and Gaussian Approximation');xlabel('Error Value');ylabel('Histogram Count');
+subplot(3,1,3);bar(Xez,Nez);hold on;plot(xvec,Gez,'r', 'LineWidth',2);
```
These 3 plots don't seem to match the example in the slides. It looks like the pdf doesn't fit the data quite right. I may be wrong though as I didn't dig into the reason why they don't match.

icolwell commented on matlab_simulation/code v1.0/ME597-4-GyroCovariance/gyroCovEstimation.m:
```diff
@@ -0,0 +1,90 @@
+%% A Matlab program to calculate the convariances from a Gyroscope Data.
```
This file looks identical to the one above under the directory "02-examples_practice/gyroCovEstimation/". Only one copy of the file should exist, perhaps this one should be removed.

icolwell commented on matlab_simulation/02-examples_practice/gyroCovEstimation/gyroCovEstimation.m:
```diff
@@ -0,0 +1,90 @@
+G_simf = SErr_Vec*sqrt(SErr_val)*randn(3,dataSet_length);
+% Plot the Filtered Data
+figure(2);clf;
+subplot(2,1,1); hold on;
```
The assignment mentioned decomposing the example code into functions. Since the plotting here takes up a lot of space and looks very similar to the plotting starting at line 39, maybe you could make a single plotting function and call it twice with different arguments.

icolwell commented on matlab_simulation/code v1.0/ME597-6-FastSLAM/newCode/callFS.m:
```diff
@@ -0,0 +1,9 @@
+clear all; close all; clc;
```
Some comments that explain the purpose of this file would be useful. Maybe I'm missing something, but I'm lost as to how this relates to the gyro covariance example.

icolwell commented on matlab_simulation/code v1.0/ME597-6-FastSLAM/newCode/ekf_FS1.m:
```diff
@@ -0,0 +1,34 @@
+function [mup,Spp,Hmu]=ekf_FS1(Xp,y,mu,S,Qi,check)
```
Again, I'm confused why fast slam stuff was updated in the pull request for issue #3 Gyro covariance estimation. The rest of the files in this pull request are all related to SLAM so I'm going to assume they are here for a reason and just carry on. 

Anyway, each function should begin with a block of comments to explain the function and inputs/outputs of the functions. Please see the MatLab style guide section on function comments:
[https://sites.google.com/site/matlabstyleguidelines/layout-documentation/comments#TOC-Function-header-comments-should-support-the-use-of-help-and-lookfor](https://sites.google.com/site/matlabstyleguidelines/layout-documentation/comments#TOC-Function-header-comments-should-support-the-use-of-help-and-lookfor)

icolwell commented on matlab_simulation/code v1.0/ME597-6-FastSLAM/newCode/ekf_FS1.m:
```diff
@@ -0,0 +1,34 @@
+                rp = sqrt((dx)^2+(dy)^2);
+                % Calculate Jacobian
+                Ht = [ dx/rp dy/rp; -dy/rp^2 dx/rp^2];
+                Spp= Ht\Qi*inv(Ht)';
```
As mentioned before the whitespace formatting is not consistent. The following files have the same issues so I won't bother to repeat this comment on each file.

icolwell commented on matlab_simulation/code v1.0/ME597-6-FastSLAM/newCode/ekf_FS1.m:
```diff
@@ -0,0 +1,34 @@
+                rp = sqrt((dx)^2+(dy)^2);
+                % Calculate Jacobian
+                Ht = [ dx/rp dy/rp; -dy/rp^2 dx/rp^2];
+                % Calculate Innovation
```
I would suggest cleaning up the whitespace here and matching the comments with the appropriate lines of code. It improves readability.

icolwell commented on matlab_simulation/code v1.0/ME597-6-FastSLAM/newCode/fastSLAM_func.m:
```diff
@@ -0,0 +1,174 @@
+function []=fastSLAM_func(PoseInitial,MotionNoise,MeasurementNoise,MapSize,FinalTime,dt,ParticleSize)
```
I don't see too many substantial differences between this file and "fastSLAM_func_p.m". I'm guessing one may simply be an older version of the other? Anyway, it would be best to only keep one of these files. If they are different on purpose then there may be a way to combine them and add an additional argument to the function to specify some difference in functionality that can be coded in.

icolwell commented on matlab_simulation/code v1.0/ME597-6-FastSLAM/newCode/fastSLAM_func.m:
```diff
@@ -0,0 +1,174 @@
+
+
+    %% Initialization Function   
+    function[timeInterval,finalTime,noFeatures,map,pose,totalParticles,muFeat,muFeatPred,covFeat,covFeatPred,newfeature,w,u,timeStamp,rmax,thmax]...
```
I don't often see multiple functions in the same .m file but I've read that it is OK as long as the other functions are sub-functions of the first function. This means the only place the sub-functions are ever called is within the first function. This looks like the case here but you may want to break a few of these out into separate files for use by other examples. For example the "raoBlackwellFilter_fs" and the "featureSearch_fs" functions look like some good generic functions that could be reused in other places.

LordBismaya committed:
All comments addressed except functionalization of the plots

LordBismaya committed:
Commit after review. Following Changes
1. Updated the MobileRoboticsSetup.m to add the 11-datasets folder
2. Removed a duplicate copy of this same file
3. Added the Info about this file in the top comments
4. Changed the titles of the error plots which were earlier the same.
5. Improved formatting
6. Separate Algo and Plot sections.
7. Changed the Error plot to match to that of the slides.

LordBismaya committed:
Commit after review. Following Changes made:
1. Updated the MobileRoboticsSetup.m to add the 11-datasets folder
2. Removed a duplicate copy of this same file
3. Added the Info about this file in the top comments
4. Changed the titles of the error plots which were earlier the same.
5. Improved formatting.
6. Separate Algo and Plot sections.
7. Changed the Error plot to match to that of the slides.
8*.Removed copies of the fastSLAM functions from this branch. Will modify in the respective branch.
-- Thanks Ian for the review.

icolwell commented on matlab_simulation/code v1.0/ME597-6-FastSLAM/newCode/callFS.m:
```diff
@@ -0,0 +1,9 @@
+MotionNoise=[0.02 0;0 0.02];
+MapSize=20;
+ParticleSize=100;
+fastSLAM_func_p(PoseInitial,MotionNoise,MeasurementNoise,MapSize,FinalTime,dt,ParticleSize);
```
While running this example the robot leaves the plot window for a portion of the simulation. The "fastSLAM_func_p" should likely be tweaked to plot a larger window into the world.

stevenwaslander commented on matlab_simulation/02-examples_practice/gyroCovEstimation/gyroCovEstimation.m:
```diff
@@ -0,0 +1,110 @@
+% ----Contributions in Ver.3.0
+% -1. Vectorized the Gyro Data.
+% -2. More Compact Code.
+% -3. Removed repeatations
```
repetitions

stevenwaslander commented on matlab_simulation/02-examples_practice/gyroCovEstimation/gyroCovEstimation.m:
```diff
@@ -0,0 +1,110 @@
+%% A Matlab program to calculate the convariances from a Gyroscope Data.
```
This is a lecture example, not a practice problem example, move to the other examples folder.

stevenwaslander commented on matlab_simulation/02-examples_practice/gyroCovEstimation/gyroCovEstimation.m:
```diff
@@ -0,0 +1,110 @@
+
+
+%% Plot Zone
+% Plot the Original data
```
Can you also plot the simulated data with the covariances derived without filtering? The comparison is instructive as to why the filtering makes sense.

stevenwaslander replied:
Will also make sense to make plotting functions for this, now that you have to make the same plots twice, once with the raw and once with the filtered data.
Great clean up of the plots and the code commenting is nice, well done.

LordBismaya replied:
Prof. Waslander, I did not understand what you meant by "Plot simulated data with the covariances". I believe I had already plotted the simulated data but perhaps, the title "Unfiltered Data" on the figure was a bit misleading. I have corrected that. Also, I believe the number on the plot, 1.2.3 and 4. are instructive of the whole process and to give an idea about the sequence of operations carried out. 

I will wait for your comment till the end of the day to incorporate any additional changes if required and create a pull request. Every other comment has been taken care of and ready to be committed.

LordBismaya committed:
spell check {repetitions}

LordBismaya committed:
New Plot Function. Moved to Lecture_Examples. Didnt make sense to parameterize the plot function as it wouldnt result in lesss clutter. Moving the whole code, as it is, to a new function made more sense.

LordBismaya committed:
Final Commit

Commit:
Cleaned up gyro_cov a little.

Commit:
Resolved merge conflict in MobileRoboticsSetup.m
